deployer/helm_upgrade_decision.py

```diff
@@ -0,0 +1,358 @@
+test_env = os.getenv("RUN_ENV", False)
+
+
+def discover_modified_common_files(modified_paths: list):
+    """There are certain common files which, if modified, we should upgrade all hubs
+    and/or all clusters appropriately. These common files include the helm charts we
+    deploy, as well as the GitHub Actions and deployer package we use to deploy them.
+    Args:
+        modified_paths (list[str]): The list of files that have been added or modified
+            in a given GitHub Pull Request.
+    Returns:
+        upgrade_all_clusters (bool): Whether or not all clusters should be upgraded
+            since the support chart has changed
+        upgrade_all_hubs (bool): Whether or not all hubs on all clusters should be
+            upgraded since a core piece of infrastructure has changed
+    """
+    # If any of the following filepaths have changed, we should update all hubs on all clusters
+    common_filepaths = [
+        "deployer/*",
+        "helm-charts/basehub/*",
+        "helm-charts/daskhub/*",
+        ".github/actions/deploy/*",
+        "requirements.txt",
+        ".github/workflows/deploy-hubs.yaml",
+    ]
+    # If this filepath has changes, we should update the support chart on all clusters
+    support_chart_filepath = "helm-charts/support/*"
+
+    # Discover if the support chart has been modified
+    support_matches = []
+    support_matches.extend(fnmatch.filter(modified_paths, support_chart_filepath))
+    upgrade_all_clusters = len(support_matches) > 0
+
+    # Discover if any common config has been modified
+    common_config_matches = []
+    for common_filepath_pattern in common_filepaths:
+        common_config_matches.extend(
+            fnmatch.filter(modified_paths, common_filepath_pattern)
+        )
+    upgrade_all_hubs = len(common_config_matches) > 0
+
+    return upgrade_all_clusters, upgrade_all_hubs
+
+
+def generate_lists_of_filepaths_and_filenames(input_file_list: list):
+    """For a list of added and modified files, generate the following:
+    - A list of unique filepaths to cluster folders containing added/modified files
+    - A set of all added/modified files matching the pattern "*/cluster.yaml"
+    - A set of all added/modified files matching the pattern "*/*.values.yaml"
+    - A set of all added/modified files matching the pattern "*/support.values.yaml"
+    Note: "cluster folders" are those that contain a cluster.yaml file.
+    Args:
+        input_file_list (list[str]): A list of files that have been added or modified
+            in a GitHub Pull Request
+    Returns:
+        list[path obj]: A list of unique filepaths to cluster folders
+        set[str]: A set of all files matching the pattern "*/cluster.yaml"
+        set[str]: A set of all files matching the pattern "*/*.values.yaml"
+        set[str]: A set of all files matching the pattern "*/*support*.values.yaml"
+    """
+    patterns_to_match = ["*/cluster.yaml", "*/*.values.yaml", "*/*support*.values.yaml"]
```

Comment:
With this, we will start relying on (but not enforcing here) the previously non-strict naming convention of having `values.yaml` in the end of all values files.

Adding this strict filter pattern makes me think we should enforce `values.yaml` suffix in `helm_chart_values_files`.

I'm brainstorming an idea that one could not need a filter pattern at all by detecting/inspecting all clusters and inspecting their helm_chart_values_yaml files etc for relevant files to observe for changes. Given how this implementation is now though, I suggest its pair with enforcement of the files listed in `helm_chart_values_files` will lead to matching the patterns defined here.

Reply:
Can you help me understand why the .values.yaml prefix is needed, vs just a .yaml?

Reply:
Actually, just ignore that question :) I don't think that's particularly important at this point, we can always re-evaluate later if needed.

Reply:
We have `helm_chart_values_files` configuration, and then we have naming conventions. With this PR we currently end up requiring both explicit configuration and following the naming convention, but we currently don't enforce the naming convention.

Overall, I appreciate if we follow a naming convention of having `values` be included in all helm chart values files because it simply clarifies something that isn't obvious. Anything including `secret` doesn't by any mean say to me that _this is a helm chart values file_ so I appreciate the verbosity and convention of including `values` in the name of a helm chart values file.

If we adopt this PRs currently logic that relies on file name patterns rather than reading the explicit configuration in `helm_chart_values_files` for each cluster to do the smart CI/CD logic, that is an acceptable compromise as long as the convention is enforced as well.

Reply:
This convention also helps us differentiate between other yaml files we have in the repo, like GitHub Actions workflow files.